control-plane/cni/main.go

```diff
@@ -135,56 +162,120 @@ func cmdAdd(args *skel.CmdArgs) error {
+	dnsIP := searchDNSIPFromEnvironment(*pod, cfg.DNSPrefix)
+	if dnsIP != "" {
+		logger.Error("assigned consul dns ip to %s", dnsIP)
+		iptablesCfg.ConsulDNSIP = dnsIP
 	}
 
-	// TODO: Get transparent proxy annotations and merge with proxy config
-	// TODO: Redirect traffic :)
+	// apply the iptables rules.
+	err = iptables.Setup(iptablesCfg)
+	if err != nil {
+		logger.Error("could not apply iptables setup: %v", err)
+		return types.PrintResult(result, cfg.CNIVersion)
+	}
 
-	// Pass through the result for the next plugin
+	logger.Debug("traffic redirect rules applied to pod: %s", pod.Name)
+	// Pass through the result for the next plugin even though we are the final plugin in the chain.
 	return types.PrintResult(result, cfg.CNIVersion)
 }
 
 // cmdDel is called for DELETE requests.
 func cmdDel(args *skel.CmdArgs) error {
-	// Nothing for consul-cni plugin to do as everything is removed once the pod is gone
+	// Nothing to do but this function will still be called as part of the CNI specification.
 	return nil
 }
 
+// cmdCheck is called for CHECK requests.
 func cmdCheck(args *skel.CmdArgs) error {
+	// Nothing to do but this function will still be called as part of the CNI specification.
 	return nil
 }
 
 func main() {
 	skel.PluginMain(cmdAdd, cmdCheck, cmdDel, version.All, bv.BuildString("consul-cni"))
 }
 
-func hasBeenInjected(pod corev1.Pod) bool {
-	if anno, ok := pod.Annotations[keyInjectStatus]; ok && anno == injected {
+// skipConnectInject determines if the connect-inject annotation is false.
+func skipConnectInject(pod corev1.Pod) bool {
+	if anno, ok := pod.Annotations[annotationInject]; ok && anno == "false" {
 		return true
 	}
 	return false
 }
+
+// waitForAnnotation waits for an annotation to be available. Returns immediately if the annotation exists.
+func waitForAnnotation(pod corev1.Pod, annotation string, retries uint64) bool {
+	var err error
+	err = backoff.Retry(func() error {
+		var ok bool
+		_, ok = pod.Annotations[annotation]
+		if !ok {
+			return fmt.Errorf("annotation %s does not exist yet", annotation)
+		}
+		return err
+	}, backoff.WithMaxRetries(backoff.NewConstantBackOff(1*time.Second), retries))
+	return err == nil
+}
+
+// parseAnnotation parses the cni-proxy-config annotation into an iptables.Config object.
+func parseAnnotation(pod corev1.Pod, annotation string) (iptables.Config, error) {
+	anno, ok := pod.Annotations[annotation]
+	if !ok {
+		return iptables.Config{}, fmt.Errorf("could not find %s annotation for %s pod", annotation, pod.Name)
+	}
+	cfg := iptables.Config{}
+	err := json.Unmarshal([]byte(anno), &cfg)
+	if err != nil {
+		return iptables.Config{}, fmt.Errorf("could not unmarshal %s annotation for %s pod", annotation, pod.Name)
+	}
+	return cfg, nil
+}
+
+// searchDNSIPFromEnvironment gets the consul server DNS IP from the pods environment variables. The prefix makes searching easier
+// return an empty string.
+func searchDNSIPFromEnvironment(pod corev1.Pod, prefix string) string {
```

Comment:
I think we need a different way to do this. The env var will only be available if the DNS service and the application are running in the same namespace (https://kubernetes.io/docs/concepts/containers/container-environment/)

Reply:
@ishustava I was following how connect-inject does it in [container_init.go](https://github.com/hashicorp/consul-k8s/blob/main/control-plane/connect-inject/container_init.go#L148url)

I am assuming there that it is running in the context of a container so the environment variable is set already...?

Reply:
The code you linked to is executed in the webhook, meaning the webhook will do the searching of the dns service env variable, not at the time when the container-init container runs. I think we have two options:

1. Figure out the dns ip in the cni installer and pass it down to cni. The problem with this is that it won't work if IP changes (ie. kube service gets recreated)
2. Do dns lookup of the consul dns service in this function and pass the dns service name from the installer.

Reply:
This has been updated in several places. The helm chart passes the `consul-fullname`.`Release.Namespace` to the installers command line as a `-dns-service` argument. That ends up in the config, the plugin grabs it and gets the IP address from Kubernetes on every run.

Reply:
This has been removed and will be addressed by getting the environment variable in the webhook.